source/ConditionSet.cpp

```diff
@@ -40,15 +42,183 @@ namespace {
 			{"<=", [](int64_t a, int64_t b) -> int64_t { return a <= b; }},
 			{">=", [](int64_t a, int64_t b) -> int64_t { return a >= b; }},
 			{"=", [](int64_t a, int64_t b) { return b; }},
+			{"*=", [](int64_t a, int64_t b) { return a * b; }},
 			{"+=", [](int64_t a, int64_t b) { return a + b; }},
 			{"-=", [](int64_t a, int64_t b) { return a - b; }},
+			{"/=", [](int64_t a, int64_t b) { return b ? a / b : numeric_limits<int64_t>::max(); }},
 			{"<?=", [](int64_t a, int64_t b) { return min(a, b); }},
-			{">?=", [](int64_t a, int64_t b) { return max(a, b); }}
+			{">?=", [](int64_t a, int64_t b) { return max(a, b); }},
+			{"%", [](int64_t a, int64_t b) { return (a % b); }},
+			{"*", [](int64_t a, int64_t b) { return a * b; }},
+			{"+", [](int64_t a, int64_t b) { return a + b; }},
+			{"-", [](int64_t a, int64_t b) { return a - b; }},
+			{"/", [](int64_t a, int64_t b) { return b ? a / b : numeric_limits<int64_t>::max(); }}
 		};
 		
 		auto it = opMap.find(op);
 		return (it != opMap.end() ? it->second : nullptr);
 	}
+	
+	// Indicate if the operation is a comparison or modifies the condition.
+	bool IsComparison(const string &op)
+	{
+		static const set<string> comparison = {
+			"==", "!=", "<", ">", "<=", ">="
+		};
+		return comparison.count(op);
+	}
+	
+	bool IsAssignment(const string &op)
+	{
+		static const set<string> assignment = {
+			"=", "+=", "-=", "*=", "/=", "<?=", ">?="
+		};
+		return assignment.count(op);
+	}
+	
+	bool IsSimple(const string &op)
+	{
+		static const set<string> simple = {
+			"(", ")", "+", "-", "*", "/", "%"
+		};
+		return simple.count(op);
+	}
+	
+	int Precedence(const string &op)
+	{
+		static const map<string, int> precedence = {
+			{"(", 0}, {")", 0},
+			{"+", 1}, {"-", 1},
+			{"*", 2}, {"/", 2}, {"%", 2}
+		};
+		return precedence.at(op);
+	}
+	
+	// Test to determine if unsupported operations are requested.
+	bool HasInvalidOperators(const vector<string> &tokens)
+	{
+		static const set<string> invalids = {
+			"{", "}", "[", "]", "|", "^", "&", "!", "~",
+			"||", "&&", "&=", "|=", "<<", ">>"
+		};
+		for(const string &str : tokens)
+			if(invalids.count(str))
+				return true;
+		return false;
+	}
+	
+	// Ensure the ConditionSet line has balanced parentheses on both sides.
+	bool HasUnbalancedParentheses(const vector<string> &tokens)
```

Comment:
Is it possible to fully parse the expression (each and every token, including variables and sub-expressions) using a parsing technique like for example https://en.wikipedia.org/wiki/Recursive_descent_parser#C_implementation?

Your implementation supports quite some different expressions, but if such full parsing is possible, then I expect that the game can support a lot more complex expressions while at the same time also providing a stronger verification of the correctness of the expressions.

Reply:
I'm still quite green when it comes to algorithms so if a more complex parser implementation is preferred I'll have to ask you propose the changes. You can either open a PR to this branch (on my fork) or a similar PR to `master` here.

Reply:
I did some work on parsers a very long time ago, so I'm willing to give it a try.

What I usually do for complex changes is to split them into multiple smaller PRs and only start working on the next PR when the previous one is tested, reviewed and merged. For example:
* PR #4391 prepares the code, data-files and savegames for additional bay-types, but the actual change to enable different bay-types will be a separate PR.
* PR #4370 prepares the framework for auto-testing, but I have a list of about 7 different PRs that I should create to get to really good auto-tests.
Would this way of working also work for Endless Sky? Or does the merge-process require a different way of working?